src/toxcore/onion/client/mod.rs

```diff
@@ -0,0 +1,2123 @@
+                    node
+                } else {
+                    break
+                };
+
+                let path = if let Some(path) = paths_pool.random_path(friend_pk.is_some()) {
+                    path
+                } else {
+                    break
+                };
+
+                let request_id = announce_requests.new_ping_id(AnnounceRequestData {
+                    pk: node.pk,
+                    saddr: node.saddr,
+                    path_id: path.id(),
+                    friend_pk,
+                });
+
+                let inner_announce_request = announce_packet_data.request(&node.pk, None, request_id);
+                let onion_request = path.create_onion_request(node.saddr, InnerOnionRequest::InnerOnionAnnounceRequest(inner_announce_request));
+
+                packets.push((Packet::OnionRequest0(onion_request), path.nodes[0].saddr));
+            }
+        }
+
+        packets
+    }
+
+    /// Announce ourselves periodically.
+    fn announce_loop(&self, state: &mut OnionClientState) -> impl Future<Item = (), Error = RunError> + Send {
+        let announce_packet_data = AnnouncePacketData {
+            packet_sk: &self.real_sk,
+            packet_pk: self.real_pk,
+            search_pk: self.real_pk,
+            data_pk: Some(self.data_pk),
+        };
+
+        let packets = OnionClient::ping_close_nodes(
+            &mut state.announce_list,
+            &mut state.paths_pool,
+            &mut state.announce_requests,
+            announce_packet_data,
+            None,
+            None,
+        );
+
+        send_all_to(&self.dht.tx, stream::iter_ok(packets))
+            .map_err(|e| e.context(RunErrorKind::SendTo).into())
+    }
+
+    /// Get nodes to include to DHT `PublicKey` announcement packet.
+    fn dht_pk_nodes(&self) -> Vec<TcpUdpPackedNode> {
+        let relays = self.tcp_connections.get_random_relays(2);
+        let close_nodes: Vec<PackedNode> = self.dht.get_closest(&self.dht.pk, 4 - relays.len() as u8, false).into();
+        relays.into_iter().map(|node| TcpUdpPackedNode {
+            pk: node.pk,
+            ip_port: IpPort::from_tcp_saddr(node.saddr),
+        }).chain(close_nodes.into_iter().map(|node| TcpUdpPackedNode {
+            pk: node.pk,
+            ip_port: IpPort::from_udp_saddr(node.saddr),
+        })).collect()
+    }
+
+    /// Announce our DHT `PublicKey` to a friend via onion.
+    fn send_dht_pk_onion(&self, friend: &mut OnionFriend, paths_pool: &mut PathsPool) -> Vec<(Packet, SocketAddr)> {
+        let dht_pk_announce = DhtPkAnnouncePayload::new(self.dht.pk, self.dht_pk_nodes());
+        let inner_payload = OnionDataResponseInnerPayload::DhtPkAnnounce(dht_pk_announce);
+        let nonce = gen_nonce();
+        let payload = OnionDataResponsePayload::new(&precompute(&friend.real_pk, &self.real_sk), self.real_pk, &nonce, &inner_payload);
+
+        let mut packets = Vec::new();
+
+        for node in friend.close_nodes.iter() {
```

Comment:
How about filter|map|flat_map?

Reply:
It will be:
```rust
        let packets = friend.close_nodes.iter()
            .filter(|node| !node.is_timed_out())
            .flat_map(|node| node.data_pk.map(|data_pk| (node, data_pk)))
            .flat_map(|(node, data_pk)| paths_pool.use_path(node.path_id, true).map(|path| (node, data_pk, path)))
            .map(|(node, data_pk, path)| {
                let (temporary_pk, temporary_sk) = gen_keypair();
                let inner_data_request = InnerOnionDataRequest::new(&precompute(&data_pk, &temporary_sk), friend.real_pk, temporary_pk, nonce, &payload);

                let onion_request = path.create_onion_request(node.saddr, InnerOnionRequest::InnerOnionDataRequest(inner_data_request));

                (Packet::OnionRequest0(onion_request), path.nodes[0].saddr)
            })
            .collect::<Vec<_>>();
```
Might be difficult for understanding. Do you thing this is better?

Reply:
No. Keep yours